### ISSUE TITLE
New rule: F63 - link-context 

### ISSUE THREAD

**Issue by jmucha** [dequelabs/axe-core#532]
#  F63 - link-context 

This test verifies whether the purpose of the link can be determined by its context and if the purpose is not ambiguous.

**Tags:** wcag2a, wcag244, cat.semantics, experimental

## Selector

Select each node with an anchor `a[href]` or `role=link`  .

### semantics/link-context-p

- Find the context by looking for the closest parent/ancestor `p` element
- Return true it the context element does not contain another link with the same link text

### semantics/link-context-li

- Find the context by looking for the closest parent/ancestor `li` element
- Return true it the context element does not contain another link with the same link text

### semantics/link-context-table-cell 

- Find the context by looking for the closest parent/ancestor `td` or `th` element
- Return true it the context element does not contain another link with the same link text

### semantics/link-context-header

- Find the context by looking for the closest preceding header
- Return true it the context element does not contain another link with the same link text

### semantics/link-context-link-role 
- Check that links with the same text don't have the same context elements. If true, pass; else fail.

### semantics/link-context-aria
- Check if the link context can be programmatically determined by using a WAI-ARIA property such as `aria-label`, `aria-labelledby` or `aria-describedby` on the link. 
- Return true if found, else fail.

**Comment by WilcoFiers** [comment]
Couple of suggestions:

1. Make sure you select `role=link` and `a[href]`. Just `a` will catch things we don't want to test

2. Break each `context` up in its own check, this will make the rule much more manageable.

3.  The check doesn't really explain what will happen when the context is identified. I'm assuming you'll find each context element and check that links with the same text don't also have the same context elements. Is that correct? Please clarify.

4. I think `link-context` would be a good rule ID here.

5. This rule should tagged as experimental.

**Comment by jmucha** [comment]
Small update:
2. I've broken each context in the separate check. It will be indeed easier to maintain. Good point.
3. That was my intention. Having found a context element, we check whether links with the same text don't have the same context element.
5. Should we keep this rule only experimental or can it belong as well to the cat.semantics?

**Comment by WilcoFiers** [comment]
- experimental is just an additional tag. 
- The selector should be: `a[href]` or `[role=link]`
- As for breaking it up in checks, I meant for it to look like this:

```md
## semantics/link-context-p ( any )

- Find the context by looking for the closest parent/ancestor `p` element
- Return true it the context element does not contain another link with the same link text

## semantics/link-context-li ( any )

- Find the context by looking for the closest parent/ancestor `li` element
- Return true it the context element does not contain another link with the same link text

## semantics/link-context-table-cell ( any )

- Find the context by looking for the closest parent/ancestor `td` or `th` element
- Return true it the context element does not contain another link with the same link text

... etc
```

**Comment by jmucha** [comment]
Tagged as an experimental rule. 

We'll target `a[href]` or `[role=link]` elements.

Split up the checks

**Comment by jkodu** [comment]
@Danidude - 
This rule does not overlap with https://github.com/dequelabs/axe-core/issues/537, this one is checking for context. If there are any specific Q's please list them.

**Comment by Danidude** [comment]
So in theory can both this rule and the link-text rule (#537) fail because the tested element(node) contain the same link text? Or only link-text rule as the other element(node) is not within the testing area this rule covers?

And this rule only tells if the violation is because the context is ambiguous and the other that the link text is to similar?

**Comment by Danidude** [comment]
@jkodu Okay, thanks, helped a lot :)

**Comment by Danidude** [comment]
`semantics/link-context-link-role`
Unsure what this check is looking for. Is it checking if the two links do not have the same parent element? Any help with this check would be wonderful.

**Comment by Danidude** [comment]
The check `semantics/link-context-header` unsure how complex this test is going to be. Is it only checking if the header element contains an link with the same link text or check all links between the header and the next header to see if they have the same link text?

**Comment by jkodu** [comment]
@Danidude 
Following up on previous discussion. All of the above checks should be `any` and does not need an `after` check.

**Comment by WilcoFiers** [comment]
Closing the issue due to inactivity.
